### PR TITLE
add swapped event to registrar

### PR DESCRIPTION
**Pull request by brunopgalvao**
This PR adds a `Swapped` event to the `paras_registrar`.

This can be helpful to let others know that a swap has occurred.

### CODE REVIEW

**Review by bkchr — APPROVED** [review]
You could check for this event in the tests

**Comment by brunopgalvao** [comment]
> You could check for this event in the tests

I see many examples of verifying events in benchmarking but not in tests. Do we have an example that I can take a look at?

**Comment by ggwpez** [comment]
> I see many examples of verifying events in benchmarking but not in tests. Do we have an example that I can take a look at?

Yea tons of it. You can search the code-base for `assert_last_event`.

**Review by shawntabrizi — APPROVED** [review]
if you need help writing a test, let me know.

**Comment by brunopgalvao** [comment]
@shawntabrizi that would be great. I committed some code - I will leave some comments.. Any feedback would be appreciated.

File: runtime/common/src/integration_tests.rs [hunk 1]
```diff
@@ -1054,6 +1054,13 @@ fn parachain_swap_works() {
 			ParaId::from(2001),
 			ParaId::from(2000)
 		));
+		assert!(contains_event(
+            paras_registrar::Event::<Test>::Swapped {
+                para_id: ParaId::from(2001),
+                other_id: ParaId::from(2000)
+            }
+            .into()
+        ));
```
Comment (brunopgalvao) [inline runtime/common/src/integration_tests.rs:1063]: Add a check for the Swapped Event in the integration test.

File: runtime/common/src/paras_registrar.rs [hunk 2]
```diff
@@ -584,6 +586,7 @@ impl<T: Config> Pallet<T> {
 		let res2 = runtime_parachains::schedule_parathread_upgrade::<T>(to_upgrade);
 		debug_assert!(res2.is_ok());
 		T::OnSwap::on_swap(to_upgrade, to_downgrade);
```
Comment (brunopgalvao) [inline runtime/common/src/paras_registrar.rs:588]: What is the reason we switched the order? The input params are:
```rust
fn do_thread_and_chain_swap(to_downgrade: ParaId, to_upgrade: ParaId)
```
but the params are switched here:
```rust
T::OnSwap::on_swap(to_upgrade, to_downgrade);
```
Why not keep it the same.

File: runtime/common/src/paras_registrar.rs [hunk 3]
```diff
@@ -584,6 +586,7 @@ impl<T: Config> Pallet<T> {
...
+		Self::deposit_event(Event::<T>::Swapped { para_id: to_upgrade, other_id: to_downgrade });
```
Comment (brunopgalvao) [inline runtime/common/src/paras_registrar.rs:589]: This was throwing me off a bit. Should it be:
```rust
Swapped { para_id: to_upgrade, other_id: to_downgrade }
```
OR
```rust
Swapped { para_id: to_downgrade, other_id: to_upgrade }
```

File: runtime/common/src/paras_registrar.rs [hunk 4]
```diff
@@ -1052,9 +1055,21 @@ mod tests {
...
+                System::events(),
+                vec![EventRecord {
+                    phase: Phase::Initialization,
+                    event: Event::Registrar(paras_registrar::Event::Swapped {
+                        para_id: para_2,
+                        other_id: para_1
+                    }),
+                    topics: vec![],
+                }]
+            );
```
Comment (brunopgalvao) [inline runtime/common/src/paras_registrar.rs:1072]: I could have written the test this way:
```rust
assert!(System::events().iter().any(|r| matches!(
        r.event,
        Event::Registrar(paras_registrar::Event::Swapped { .. })
    )));
```
But in this way I was not able to check for the properties inside the Swapped Event.

File: runtime/common/src/paras_registrar.rs [hunk 5]
```diff
@@ -1067,6 +1082,28 @@ mod tests {
...
+                    phase: Phase::Initialization,
+                    event: Event::Registrar(paras_registrar::Event::Swapped {
+                        para_id: para_1,
+                        other_id: para_2
+                    }),
+                    topics: vec![],
+                }]
+                .last()
+                .unwrap()
+            );
```
Comment (brunopgalvao) [inline runtime/common/src/paras_registrar.rs:1102]: We did not have a test case for a swap between parathread and parachain which is one of the `else if` conditions here:
- https://github.com/paritytech/polkadot/blob/53cc5146bf505fbb2a110a815404703304df2c8e/runtime/common/src/paras_registrar.rs#L297-L300

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2577341

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2657114

File: runtime/common/src/paras_registrar.rs (hunk 2 above)
Comment (bkchr) [reply runtime/common/src/paras_registrar.rs:651]: Looks like a bug?

Comment (brunopgalvao) [reply runtime/common/src/paras_registrar.rs]: I do not believe it was a bug, just a way to reuse `do_thread_and_chain_swap()` for parathread and parachain and vice versa

**Comment by brunopgalvao** [comment]
Will solve issue https://github.com/paritytech/polkadot/issues/5562

**Review by liamaharon — APPROVED** [review]

**Comment by brunopgalvao** [comment]
bot merge

**Pull request merged**